### ISSUE TITLE
`no relation` `label` data는 없다고 생각하자

### ISSUE THREAD

**Issue by yhkee0404** [BoocampMonster/korean-relation-extraction#14]
주어진 문제는 다음 두 가지 경우밖에 없다. 첫째, 정답이 `no relation`인 경우와, 둘째, 정답이 `29`개의 `yes relation` 중 하나인 경우이다. 이때 본 대회는 정답이 `no relation`인 경우를 제외한 점수 우선으로 (`preemptively`) 채점한다는 특징이 있다. 이 글에서는 해당 점수만 고려한다.

 첫째, 정답이 `no relation`인 경우, 예측으로 `no relation`을 맞히거나 틀리거나 득점은 `0`점으로 동일하다. 즉 예측을 무조건 `yes relation` 중 하나로 해서 항상 오답을 출력해도 아무런 문제가 없다.

 둘째, 정답이 `29`개의 `yes relation` 중 하나인 경우, 그 특정 `label`을 가령 `per:children`이라고 하자. 이때 예측을 `per:children`이라고 맞히면 득점을 하지만 다른 `label`을 출력하면 감점을 받는다. 그러면 오답으로 감점을 받는 때를 다음과 같이 세부적인 두 가지 경우로 나눌 수 있다. 각 경우를 `2-A`와 `2-B`라고 부르자.

`2-A`: 예측하는 `label`이 `no relation`을 포함한 `30`개 중 하나인 경우, 감점 받는 경우는 정답 `per:children`을 제외한 `29`개이다. 일반적으로 `no relation`으로 예측할 확률이 `0%`보다 클수록, 비둘기 집의 원리에 따라 정답 `per:children`을 맞힐 확률은 같거나 줄어든다. 다시 말해 감점의 위험이 커진다.

`2-B`: 예측하는 `label`이 `no relation`을 제외한 `29`개 중 하나인 경우, 감점 받는 경우는 정답 `per:children`을 제외한 `28`개이다. 이때 예측으로 정답을 맞힐 확률은 `2-A`에서 `no relation`으로 예측할 확률이 `0%`인 경우와 같거나 더 커진다. 따라서 `2-B`에서 감점의 위험은 `2-A`에서 언급한 감점 위험의 최소한과 같거나 더 작다. 즉 `2-B`가 `2-A`보다 결코 불리한 경우는 없으면서 일반적으로 더 유리하다.

 끝으로, 첫째, 둘째, 둘째 안에서는 `2-A`, `2-B`로 나눈 각각의 경우를 모두 종합하면, 본 대회에서 주어진 문제는 항상 `no relation` `label` data가 없다고 생각하고 예측할 때가 그러지 않을 때보다 결코 불리한 경우는 없으면서 일반적으로 더 유리하다. 그래서 일반적인 `30`개 분류기 학습이나, 기존에 추진할 예정이었던 `no relation`과 `yes relation`의 이진 (`binary`) 분류기 학습 계획 조차도 포기하고, 오직 `29`개 분류기만 학습할 것을 제안한다.

 덧붙여, 그러면 기존에 독보적으로 골치를 썩이던 `no relation` `label` data만 유독 과다하다는 데이터 불균형 (`imbalance`) 문제까지 일부 덩달아 해결 되니 더욱 잘 됐다. 반박 환영한다.

 본 `issue`는 다음 `issuecomment`에서 확장한 것이다: 
__Originally posted by @yhkee0404 in https://github.com/BoocampMonster/korean-relation-extraction/issues/4#issuecomment-1328022470__

**Comment by hundredeuk2** [comment]
좋은 의견 감사합니다! 확실히 이렇게 하니까 재밌네요 ㅋㅋㅋㅋㅋㅋ
우선 경우의 수를 함께 해주셔서 이해하기 편하네요. 이건 무조건 부정이라기 보단 모델이 추론을 해야하는 행위에 대한 관점에서의 해석과 의문임을 밝히고 들어가겠습니다!

1. 또다시 Metric 관점에선 이해했습니다. 하지만 Loss의 개념에선  저희가 Backpropbagation을 통해 학습하기에 Loss값을 계산합니다. 그럼 no relation인 경우에도 맞춰야만 loss값이 줄어들면서 학습을 하겠죠. "아 이건 확실히 no relation"이다. 그것이 아니라면 정답은 no_relation이 아닌데 no_relation이라고 판단한 부분에 대해서는 학습이 안되지 않을까요?

다음은 ignore index = 0 을 하고 전체 class를 나타내봤습니다.
<img width="607" alt="image" src="https://user-images.githubusercontent.com/73874591/204124087-a754acbd-0f1c-4283-b735-2a336b07d21e.png">

2. label의 특수함 때문이라고 생각합니다. #4 이슈에 커맨트에 따라 1~9번 label만 쓰기위해 나머지 20개의 인덱스를 무시한건 이미지가 Captioning한 Area에 무조건 사람만 있다고 가정을 할 수 없기 때문입니다. 풍경과 사람이 같이 찍히는 Boxline이 있을 경우 이것에 대한 예측 분포를 Person / Tree 등으로 할 수 있으니깐요, 그런데 No relation이면서 per:children일 수 있을까요? 저는 우선 아니라고 생각합니다. 그렇기 때문에 둘 중 하나를 먼저 선별하는 이진 분류기로 분류를 시작하고 넘어가면 Loss기준으로 29개의 Label에 Focus 할 수 있다는 생각입니다.
3. no relation을 제외한 29개든 28개든 그거랑은 상관 없이 그 선택지 자체를 훈련과 확률에 의해 먼저 걸러주고 들어가자는게 제 아이디어의 핵심입니다. 왜냐하면 데이터의 분포가 no relation이 절반을 차지하기 때문에 no relation으로 예측할 확률을 계산조차 안 하겠다는 것입니다.
4. 이렇게 되면 Label에 대한 정보가 전혀 없는 Test Dataset에 대해서는 no relation을 어떻게 주려고 제안하신건지 의문입니다.
5. 그래서 결론이 무엇인지가 궁금합니다. 말씀해주신건 Val_Loss를 no relation을 포함 안할 때, 정답이 no relation만 거르고 예측이 no relation일 때는 넣어주자는 의견 아니였는지

**Comment by hundredeuk2** [comment]
제가 4번 이슈를 같이 읽어서 난독증이 잠깐 있었나봐요 ㅋㅋㅋㅋㅋㅋㅋㅋㅋ no relation을 학습하지 않은 분류기에 그냥 넣자는 이야기군요. 조금은 편법같지만 이해했습니다

**Comment by yhkee0404** [comment]
하지만 실험 결과 내가 틀렸다 ㅜㅜ 우리 대회 `Micro F1` 점수를 잘못 이해한 것 같아요. 예측 오답으로 `no relation`을 제시하는 것이 다른 오답 가령 `per:sibling`을 제시하는 것보다 더 유리해서 그런 것 같아. 추측해 보면 전자 오답 `no relation`의 `False Positive`는 (실제 `per:children`, 예측 `no relation`) 계산에서 제외 되는 것이 맞지만 후자 오답 `per:sibling`의 `False Positive`는 (실제 `per:children`, 예측 `per:sibling`) 계산에 포함 돼서 감점이 커지는 것 같다.
